### PR TITLE
CAMEL-12005: Add websocket support to camel-undertow

### PR DESCRIPTION
**Pull request by ppalaga**

### CODE REVIEW

**Comment by oscerd** [comment]
This looks good. @davsclaus wdyt?

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConstants.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.undertow;
+
+/**
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
```
Comment (davsclaus) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConstants.java]: Remove author tags in javadoc

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowEndpoint.java [hunk 2]
```diff
@@ -87,6 +92,14 @@
     private boolean optionsEnabled;
     @UriParam(label = "producer")
     private CookieHandler cookieHandler;
+    @UriParam(label = "producer")
```
Comment (davsclaus) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowEndpoint.java]: Add websocket as 2nd label, eg "producer,websocket" and do that for the other new websocket options

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowProducer.java [hunk 3]
```diff
@@ -77,81 +83,135 @@ public UndertowEndpoint getEndpoint() {
...
+                throw new IllegalStateException(
+                        String.format("Cannot process message which has none of the headers %s, %s or %s set: %s",
+                                UndertowConstants.SEND_TO_ALL, UndertowConstants.CONNECTION_KEY_LIST,
+                                UndertowConstants.CONNECTION_KEY, in));
+            } else {
+                /* nothing to do for a null body */
+                camelCallback.done(true);
+                return true;
+            }
+        } catch (IOException e) {
```
Comment (davsclaus) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowProducer.java]: Catch all kinds of exceptions, eg Exception or Throwable

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/CamelMethodHandler.java [hunk 4]
```diff
@@ -16,76 +16,127 @@
...
     }
 
-    public boolean isEmpty() {
-        return defaultHandler == null && methodMap.isEmpty();
-    }
+    static class MethodEntry {
+        /**
+         * The number of references pointing to {@link #handler}
+         */
+        private int refCount;
```
Comment (davsclaus) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/CamelMethodHandler.java]: Add a bit of empty lines so this code is formatted nice between fields, constructor, and methods

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+package org.apache.camel.component.undertow.handlers;
+
+import io.undertow.websockets.core.WebSocketCallback;
+import io.undertow.websockets.core.WebSocketChannel;
+
+/**
+ * An extension of {@link WebSocketCallback} that adds a notification for the situation when a {@link WebSocketChannel}
+ * is closed before any message could be sent to it.
+ *
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
```
Comment (davsclaus) [inline components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java]: Remove author tags, check other files where you may have that too

File: components/camel-undertow/src/test/java/org/apache/camel/component/undertow/ws/UndertowWssRouteTest.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        AsyncHttpClient c = createAsyncHttpSSLClient();
+        WebSocket websocket = c.prepareGet("wss://localhost:" + getPort() + "/test").execute(
+                new WebSocketUpgradeHandler.Builder()
+                        .addWebSocketListener(new WebSocketTextListener() {
+                            @Override
+                            public void onMessage(String message) {
+                                received.add(message);
+                                log.info("received --> " + message);
+                                latch.countDown();
+                            }
```
Comment (davsclaus) [inline components/camel-undertow/src/test/java/org/apache/camel/component/undertow/ws/UndertowWssRouteTest.java]: Remove all the extra empty lines below

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java (hunk 5 above)
Comment (ppalaga) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java]: > Remove author tags

Why if I may ask?

Comment (davsclaus) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java]: That is ASF policy

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConstants.java (hunk 1 above)
Comment (ppalaga) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConstants.java]: Fixed in a85b3e3

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowEndpoint.java (hunk 2 above)
Comment (ppalaga) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowEndpoint.java]: Fixed in a85b3e3

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowProducer.java (hunk 3 above)
Comment (ppalaga) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowProducer.java]: Fixed in a85b3e3

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/CamelMethodHandler.java (hunk 4 above)
Comment (ppalaga) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/CamelMethodHandler.java]: Fixed in a85b3e3

File: components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java (hunk 5 above)
Comment (ppalaga) [reply components/camel-undertow/src/main/java/org/apache/camel/component/undertow/handlers/ExtendedWebSocketCallback.java]: Fixed in a85b3e3

File: components/camel-undertow/src/test/java/org/apache/camel/component/undertow/ws/UndertowWssRouteTest.java (hunk 6 above)
Comment (ppalaga) [reply components/camel-undertow/src/test/java/org/apache/camel/component/undertow/ws/UndertowWssRouteTest.java]: Fixed in a85b3e3

**Comment by ppalaga** [comment]
@davsclaus I addressed all your points in a85b3e3

**Comment by oscerd** [comment]
Merged on master. Thanks for the PR. Closing this.

**Pull request closed without merging**